cli/commands.c

```diff
@@ -0,0 +1,230 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "glome.h"
+
+#define GLOME_CLI_MAX_MESSAGE_LENGTH 4095
+
+// Arguments
+static const char *key_file = NULL;
+static const char *peer_file = NULL;
+static const char *tag_hex = NULL;
+static unsigned long counter = 0;
+
+static bool parse_args(int argc, char **argv) {
+  struct option long_options[] = {{"key", required_argument, 0, 'k'},
+                                  {"peer", required_argument, 0, 'p'},
+                                  {"counter", required_argument, 0, 'c'},
+                                  {"tag", required_argument, 0, 't'},
+                                  {0, 0, 0, 0}};
+
+  while (1) {
+    switch (getopt_long(argc, argv, "c:k:p:t:", long_options, NULL)) {
+      case -1:
+        return true;
+      case 'c': {
+        char *endptr;
+        counter = strtoul(optarg, &endptr, 0);
+        if (counter > UCHAR_MAX || optarg[0] == '\0' || *endptr != '\0') {
+          fprintf(stderr, "'%s' is not a valid counter (0..255)\n", optarg);
+          return false;
+        }
+        break;
+      }
+      case 'k':
+        key_file = optarg;
+        break;
+      case 'p':
+        peer_file = optarg;
+        break;
+      case 't':
+        tag_hex = optarg;
+        break;
+      case '?':
+        return false;
+      default:
+        // option not implemented
+        abort();
+    }
+  }
+  return true;
+}
+
+static int decode_hex(uint8_t *dst, size_t dst_len, const char *in) {
+  size_t len = strlen(in);
+  if (len > 2 && in[0] == '0' && in[1] == 'x') {
+    len -= 2;
+    in += 2;
+  }
+  if (len > dst_len * 2 || len % 2 != 0) {
+    return -1;
+  }
+  size_t i;
+  for (i = 0; i < len / 2; i++) {
+    if (sscanf(in + (i * 2), "%02hhX", dst + i) != 1) {
+      fprintf(stderr, "ERROR while parsing byte %zu ('%c%c') as hex\n", i,
+              in[2 * i], in[2 * i + 1]);
+      return -3;
+    }
+  }
+  return i;
+}
+
+static bool read_file(const char *fname, uint8_t *buf, const size_t num_bytes) {
+  FILE *f = fopen(fname, "r");
+  if (!f) {
+    fprintf(stderr, "could not open file %s: %s\n", fname, strerror(errno));
+    return false;
+  }
+  if (fread(buf, 1, num_bytes, f) != num_bytes) {
+    fprintf(stderr, "could not read %zu bytes from file %s", num_bytes, fname);
+    if (ferror(f)) {
+      fprintf(stderr, ": %s\n", strerror(errno));
+    } else {
+      fputs("\n", stderr);
+    }
+    fclose(f);
+    return false;
+  }
+  return true;
+}
+
+static void print_hex(FILE *stream, const char *prefix, uint8_t *buf,
+                      size_t len) {
+  if (prefix != NULL) {
+    fputs(prefix, stream);
+  }
+  for (size_t i = 0; i < len; i++) {
+    fprintf(stream, "%02x", buf[i]);
+  }
+  fputs("\n", stream);
+}
+
+int genkey(int argc, char **argv) {
+  uint8_t private_key[GLOME_MAX_PRIVATE_KEY_LENGTH] = {0};
+
+  // TODO: maybe verify stdout umask 077
```

Comment:
What does that mean?

Reply:
In theory, we could verify that the file associated with stdout - *if there is a file* - is not world readable. wireguard does that. However, this does not help detecting an insecure `glome genkey | tee /tmp/foo.key`, so I'm inclined to just dropping the topic.